File: scala/sdktest/OutputGetLogoEntityTest.scala
// Generated basic-flow test for the output_get_logo entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped OutputGetLogoTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefintecsuserbackofficesdk.core.{Helpers, BluefinTecsUserBackofficeSDK}
import voxgig.bluefintecsuserbackofficesdk.utility.struct.Struct

object OutputGetLogoEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("output_get_logo.instance") {
      val testsdk = BluefinTecsUserBackofficeSDK.testSDK()
      val ent = testsdk.outputGetLogo(null)
      rep.check("output_get_logo.instance", ent != null, "expected non-null output_get_logo entity")
    }

    rep.scope("output_get_logo.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/output_get_logo/OutputGetLogoTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinTecsUserBackofficeSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("output_get_logo01", "OUTPUT_GET_LOGO01")
      idmap.put("output_get_logo02", "OUTPUT_GET_LOGO02")
      idmap.put("output_get_logo03", "OUTPUT_GET_LOGO03")
      val now = System.currentTimeMillis()

      // LOAD
      val outputGetLogoRef01Ent = client.outputGetLogo(null)
      val outputGetLogoRef01MatchDt0 = new LinkedHashMap[String, Object]()
      val outputGetLogoRef01DataDt0Loaded = outputGetLogoRef01Ent.load(outputGetLogoRef01MatchDt0, null)
      rep.check("output_get_logo.load.nonnull", outputGetLogoRef01DataDt0Loaded != null, "expected load result to be non-null")
    }
  }
}

File: cpp/test/output_get_logo_entity_test.cpp
// Generated basic-flow test for the output_get_logo entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct OutputGetLogoSetup {
  std::shared_ptr<BluefinTecsUserBackofficeSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static OutputGetLogoSetup output_get_logo_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/output_get_logo/OutputGetLogoTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinTecsUserBackofficeSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("output_get_logo01"), Value("output_get_logo02"), Value("output_get_logo03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINTECSUSERBACKOFFICE_TEST_OUTPUT_GET_LOGO_ENTID", idmap},
    {"BLUEFINTECSUSERBACKOFFICE_TEST_LIVE", Value("FALSE")},
    {"BLUEFINTECSUSERBACKOFFICE_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINTECSUSERBACKOFFICE_TEST_OUTPUT_GET_LOGO_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINTECSUSERBACKOFFICE_TEST_LIVE") == Value("TRUE");

  OutputGetLogoSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void output_get_logo_entity_instance() {
  auto testsdk = BluefinTecsUserBackofficeSDK::testSDK();
  auto ent = testsdk->output_get_logo();
  ASSERT_EQ(ent->getName(), std::string("output_get_logo"), "entity name");
}

static void output_get_logo_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"output_get_logo", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinTecsUserBackofficeSDK::testSDK(seed, sdkopts);
  auto se = strsdk->output_get_logo();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinTecsUserBackofficeSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->output_get_logo();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void output_get_logo_entity_basic() {
  auto setup = output_get_logo_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"load"}) {
    auto sk = is_control_skipped("entityOp", std::string("output_get_logo.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;

  // Bootstrap entity data from existing test data (no create step in flow).
  // Declare _data at FUNCTION scope (later load/update steps reference it);
  // only _data_raw was declared, so the block-local assignment left _data
  // undeclared ("was not declared in this scope").
  Value output_get_logo_ref01_data_raw = Helpers::toMapAny(Struct::getpath(setup.data, {"existing", "output_get_logo"}));
  Value output_get_logo_ref01_data = vmap();
  {
    std::vector<Value> its = Struct::items(output_get_logo_ref01_data_raw);
    output_get_logo_ref01_data = its.empty() ? vmap() : Helpers::toMapAny(pair_val(its[0]));
    if (!output_get_logo_ref01_data.is_map()) output_get_logo_ref01_data = vmap();
  }
  // LOAD
  auto output_get_logo_ref01_ent = client->output_get_logo();
  Value output_get_logo_ref01_match_dt0 = vmap();
  Value output_get_logo_ref01_data_dt0_loaded = output_get_logo_ref01_ent->load(output_get_logo_ref01_match_dt0, Value::undef());
  ASSERT_TRUE(!output_get_logo_ref01_data_dt0_loaded.is_undef(), "expected load result to be non-nil");

}

int main() {
  T_RUN(output_get_logo_entity_instance);
  T_RUN(output_get_logo_entity_stream);
  T_RUN(output_get_logo_entity_basic);
  return sdktest::summary("output_get_logo_entity_test");
}
